man/limits.5
.\"     Title: limits
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 07/30/2006
.\"    Manual: File Formats and Conversions
.\"    Source: File Formats and Conversions
.\"
.TH "LIMITS" "5" "07/30/2006" "File Formats and Conversions" "File Formats and Conversions"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
limits \- resource limits definition
.SH "DESCRIPTION"
.PP
The
\fIlimits\fR
file (\fI/etc/limits\fR
by default or LIMITS_FILE defined
\fIconfig.h\fR) describes the resource limits you wish to impose. It should be owned by root and readable by root account only.
.PP
By default no quota is imposed on 'root'. In fact, there is no way to impose limits via this procedure to root\-equiv accounts (accounts with UID 0).
.PP
Each line describes a limit for a user in the form:
.PP

\fIuser LIMITS_STRING\fR
.PP
The
\fILIMITS_STRING\fR
is a string of a concatenated list of resource limits. Each limit consists of a letter identifier followed by a numerical limit.
.PP
The valid identifiers are:
.TP 3n
\(bu
A: max address space (KB)
.TP 3n
\(bu
C: max core file size (KB)
.TP 3n
\(bu
D: max data size (KB)
.TP 3n
\(bu
F: maximum filesize (KB)
.TP 3n
\(bu
M: max locked\-in\-memory address space (KB)
.TP 3n
\(bu
N: max number of open files
.TP 3n
\(bu
R: max resident set size (KB)
.TP 3n
\(bu
S: max stack size (KB)
.TP 3n
\(bu
T: max CPU time (MIN)
.TP 3n
\(bu
U: max number of processes
.TP 3n
\(bu
K: file creation mask, set by
\fBumask\fR(2).
.TP 3n
\(bu
L: max number of logins for this user
.TP 3n
\(bu
P: process priority, set by
\fBsetpriority\fR(2).
.sp
.RE
.PP
For example,
\fIL2D2048N5\fR
is a valid
\fILIMITS_STRING \fR. For reading convenience, the following entries are equivalent:
.sp
.RS 3n
.nf
      username L2D2048N5
      username L2 D2048 N5
    
.fi
.RE
.PP
Be aware that after
\fIusername\fR
the rest of the line is considered a limit string, thus comments are not allowed. A invalid limits string will be rejected (not considered) by the
\fBlogin\fR
program.
.PP
The default entry is denoted by username "\fI*\fR". If you have multiple
\fIdefault\fR
entries in your
\fILIMITS_FILE\fR, then the last one will be used as the default entry.
.PP
To completely disable limits for a user, a single dash "\fI\-\fR" will do.
.PP
Also, please note that all limit settings are set PER LOGIN. They are not global, nor are they permanent. Perhaps global limits will come, but for now this will have to do ;)
.SH "FILES"
.TP 3n
\fI/etc/limits\fR
.SH "SEE ALSO"
.PP

\fBlogin\fR(1),
\fBsetpriority\fR(2),
\fBsetrlimit\fR(2).


NEWS
$Id: NEWS,v 1.492 2007/02/01 20:49:25 kloczek Exp $

shadow-4.0.18.1 -> shadow-4.0.18.2					13-10-2007

*** general:
- usermod: fixed handle -a option (by Benno Schulenberg
  <bensberg@justemail.net>),
- useradd: improved auditing support
  (https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=211659),
- groupadd, groupdel, groupmod, useradd, userdel, usermod: flush nscd cashes
  after close /etc/{group,passwd} files,
- su: If compiled without PAM support, enforce the limits from /etc/limits
  when one of the -, -l, or --login options is set, even if called by root.
*** documentation:
- updated translations: ja, nl, tl.
- groupadd.8, groupmod.8, login.1, useradd.8, userdel.8, usermod.8: grammar
  mistakes and other correctstions (by Schulenberg <bensberg@justemail.net>),

shadow-4.0.18 -> shadow-4.0.18.1					03-08-2006

*** general:
- groupmems: fixed compilation when PAM is disabled
  (by Johannes Winkelmann <jw@smts.ch>),
- fixed missing man pages in dist tar ball necessary on build when
  PAM is disabled.

shadow-4.0.17 -> shadow-4.0.18						01-08-2006

*** general:
- su: fixed set enviroment too early when using PAM, so move it to !USE_PAM
  (patch submitted by Mike Frysinger <vapier@gentoo.org>),
- groupadd, groupmod, useradd, usermod: fixed UID/GID overflow (fixed
  http://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=198920)
- passwd, useradd, usermod: fixed inactive/mindays/warndays/maxdays overflow
  (simillar to RH#198920),
- groupmems: rewrited for use PAM and getopt_long() and now it is enabled
  for build and install (patch by George Kraft <gk4@swbell.net>),
- S/Key: removed assign getpass() to libshadow_getpass() on autoconf level
  (patch by Ulrich Mueller <ulm@kph.uni-mainz.de>; http://bugs.gentoo.org/139966),
- usermod: back to previous -a option semantics and clarify -a behavior
  on documentation level (by Greg Schafer <gschafer@zip.com.au>),
- chsh, groupmod: rewrited for use getopt_long().
- updated translations: ca, cs, da, eu, fr, gl, hu, ko, pl, pt, ru, sv, tr, uk, vi.
*** documentation:
- fr and ru man pages are up to date,
- partially translated sv man pages set added
  (by Daniel Nylander <info@danielnylander.se>),
- pl chage(1), chsh(1), groupmod(8): translation finished.

shadow-4.0.16 -> shadow-4.0.17						10-07-2006

*** general:
- userdel, usermod: fixed segfault on remove home directory when it can't
  be removed; for example when it is /dev/null (fixed http://bugs.gentoo.org/139148),
- improved SELinux detection on autoconf level (based on patch by
  Dan Yefimov <dan@D00M.lightwave.net.ru>),
- removed using private implementation getpass() libc function
  (now getpass() is used also when S/KEY support is enabled),
- move nologin do $(sbindir),
- useradd: fixed mail spool file creation (bug cached by Frans Pop
  <elendil@planet.nl>;
  fixed http://bugs.debian.org/374705),
- updated translations: cs, da, de, ko, nb, nl, pt, ro, ru, sk, sv, vi,
- new translations: dz, km, ne.
*** documentation:
- ru man pages up to date,
- lastlog(8): updated pl translation,
- faillog(5): added missing information about fail_locktime element of
  faillog struct (by Thorsten Kukuk <kukuk@suse.de>),
- updated translations: eu, fr, pl.
- reverted using docbook.sourceforge.net in XSL url.

shadow-4.0.15 -> shadow-4.0.16						05-06-2006

*** general:
- userdel: better fix for old CERT VU#312962 (which was fixed in shadow 4.0.8):
  fixed forgoten checking of the return value from fchown() before
  proceeding with the fchmod() (based on Owl patch prepared by
  Rafal Wojtczuk <nergal@owl.openwall.com>),
- userdel: use login.defs::MAIL_DIR instead hardcoded /var/mail in created
  mailbox path (based on Owl fixes submited
  by Solar Designer <solar@openwall.com>),
- by default do not use libshadow_getpass() as getpass() replacemement.
  Use libshadow_getpass() only when S/KEY support is enabled.
  Current glibc getpass() handles correctly longer than 8 characters
  passwords and libshadow_getpass() is used only because libc getpass()
  do not handles password prompting with echo enabled,
- move login.defs::MD5_CRYPT_ENAB to non-PAM part,
- userdel: rewrited for use getopt_log(),
- install default/template configuration files:
-- if shadow is configured with use PAM install /etc/pam.d/* files,
-- if shadow do not uses PAM install /etc/{limits,login.acces} files,
-- install /etc/login.defs and /etc/default/useradd files,
- fixed handle relative symlinks too in lib/commonio.c
  (merge patch from Fedora),
- properly notify nscd to flush its cache
  (https://bugzilla.redhat.com/bugzilla/186803),
- useradd, usermod: fixes for verify return values mkdir() and chown()
  on copy files (merge 482_libmisc_copydir_check_return_values Debian
  patch),
- login, su (non-PAM variant): export MAIL only when MAIL_CHECK_ENAB
  is enabled (Mike Frysinger <vapier@gentoo.org>),
- pgck, grpck: warn when the members of a group differ in /etc/groups
  and /etc/gshadow (fixed http://bugs.debian.org/75181),
- su: fixed exit with a status 0 when the invoked command is terminated
  by a signal which was not catched
  (fixed by Eero Häkkinen <eero17@bigfoot.com>),
- login: cancel login timeout after authentication so that patient people
  timing out on network directory services can log in with local
  accounts (http://bugs.debian.org/107148),
- chgpasswd: fixes for build correctly with --disable-shadowgrp
  (patch by Johannes Winkelmann <jw@tks6.net>).
- updated translations: cs, da, es, eu, fi, fr, gl, hu, id, pt, ru, sk, sv, vi.
- new translations: hu.
*** documentation:
- new cs man pages: groupmems(8), groupmod(8), grpck(8), gshadow(5)
  (by Miroslav Kure <kurem@upcase.inf.upol.cz>),
- regenerate roff man pages using docbook-style-xsl-1.70.1,
- bunch of cleanups in chfn(1), faillog(8), gpasswd(1), groupadd(8),
  groupmems(8), limits(5), login(1), login.defs(5), newgrp(1), passwd(1),
  passwd(5) and su(1) (by Yuri Kozlov <kozlov.y@gmail.com>),
- update pl vipw(8) man page,
- added chgpasswd(8) ru man page,

libmisc/limits.c
/*
 * Copyright 1989 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Separated from setup.c.  --marekm
 * Resource limits thanks to Cristian Gafton.
 */

#include <config.h>

#ifndef USE_PAM

#ident "$Id: limits.c,v 1.21 2006/07/10 04:11:31 kloczek Exp $"

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include "prototypes.h"
#include "defines.h"
#include <pwd.h>
#include "getdef.h"
#ifdef HAVE_SYS_RESOURCE_H
#include <sys/resource.h>
#define LIMITS
#endif
#ifdef LIMITS
#ifndef LIMITS_FILE
#define LIMITS_FILE "/etc/limits"
#endif
#define LOGIN_ERROR_RLIMIT	1
#define LOGIN_ERROR_LOGIN	2
/* Set a limit on a resource */
/*
 *	rlimit - RLIMIT_XXXX
 *	value - string value to be read
 *	multiplier - value*multiplier is the actual limit
 */
static int
setrlimit_value (unsigned int rlimit, const char *value,
		 unsigned int multiplier)
{
	struct rlimit rlim;
	long limit;
	char **endptr = (char **) &value;
	const char *value_orig = value;

	limit = strtol (value, endptr, 10);
	if (limit == 0 && value_orig == *endptr)	/* no chars read */
		return 0;
	limit *= multiplier;
	rlim.rlim_cur = limit;
	rlim.rlim_max = limit;
	if (setrlimit (rlimit, &rlim))
		return LOGIN_ERROR_RLIMIT;
	return 0;
}


static int set_prio (const char *value)
{
	int prio;
	char **endptr = (char **) &value;

	prio = strtol (value, endptr, 10);
	if ((prio == 0) && (value == *endptr))
		return 0;
	if (setpriority (PRIO_PROCESS, 0, prio))
		return LOGIN_ERROR_RLIMIT;
	return 0;
}


static int set_umask (const char *value)
{
	mode_t mask;
	char **endptr = (char **) &value;

	mask = strtol (value, endptr, 8) & 0777;
	if ((mask == 0) && (value == *endptr))
		return 0;
	umask (mask);
	return 0;
}


/* Counts the number of user logins and check against the limit */
static int check_logins (const char *name, const char *maxlogins)
{
#if HAVE_UTMPX_H
	struct utmpx *ut;
#else
	struct utmp *ut;
#endif
	unsigned int limit, count;
	char **endptr = (char **) &maxlogins;
	const char *ml_orig = maxlogins;

	limit = strtol (maxlogins, endptr, 10);
	if (limit == 0 && ml_orig == *endptr)	/* no chars read */
		return 0;

	if (limit == 0) {	/* maximum 0 logins ? */
		SYSLOG ((LOG_WARN, "No logins allowed for `%s'\n", name));
		return LOGIN_ERROR_LOGIN;
	}

	count = 0;
#if HAVE_UTMPX_H
	setutxent ();
	while ((ut = getutxent ())) {
#else
	setutent ();
	while ((ut = getutent ())) {
#endif
		if (ut->ut_type != USER_PROCESS)
			continue;
		if (ut->ut_user[0] == '\0')
			continue;
		if (strncmp (name, ut->ut_user, sizeof (ut->ut_user)) != 0)
			continue;
		if (++count > limit)
			break;
	}
#if HAVE_UTMPX_H
	endutxent ();
#else
	endutent ();
#endif
	/*
	 * This is called after setutmp(), so the number of logins counted
	 * includes the user who is currently trying to log in.
	 */
	if (count > limit) {
		SYSLOG ((LOG_WARN, "Too many logins (max %d) for %s\n",
			 limit, name));
		return LOGIN_ERROR_LOGIN;
	}
	return 0;
}

/* Function setup_user_limits - checks/set limits for the curent login
 * Original idea from Joel Katz's lshell. Ported to shadow-login
 * by Cristian Gafton - gafton@sorosis.ro
 *
 * We are passed a string of the form ('BASH' constants for ulimit)
 *     [Aa][Cc][Dd][Ff][Mm][Nn][Rr][Ss][Tt][Uu][Ll][Pp]
 *     (eg. 'C2F256D2048N5' or 'C2 F256 D2048 N5')
 * where:
 * [Aa]: a = RLIMIT_AS		max address space (KB)
 * [Cc]: c = RLIMIT_CORE	max core file size (KB)
 * [Dd]: d = RLIMIT_DATA	max data size (KB)
 * [Ff]: f = RLIMIT_FSIZE	max file size (KB)
 * [Mm]: m = RLIMIT_MEMLOCK	max locked-in-memory address space (KB)
 * [Nn]: n = RLIMIT_NOFILE	max number of open files
 * [Rr]: r = RLIMIT_RSS		max resident set size (KB)
 * [Ss]: s = RLIMIT_STACK	max stack size (KB)
 * [Tt]: t = RLIMIT_CPU		max CPU time (MIN)
 * [Uu]: u = RLIMIT_NPROC	max number of processes
 * [Kk]: k = file creation masK (umask)
 * [Ll]: l = max number of logins for this user
 * [Pp]: p = process priority -20..20 (negative = high, positive = low)
 *
 * Return value:
 *		0 = okay, of course
 *		LOGIN_ERROR_RLIMIT = error setting some RLIMIT
 *		LOGIN_ERROR_LOGIN  = error - too many logins for this user
 *
 * buf - the limits string
 * name - the username
 */
static int do_user_limits (const char *buf, const char *name)
{
	const char *pp;
	int retval = 0;

	pp = buf;

	while (*pp != '\0')
		switch (*pp++) {
#ifdef RLIMIT_AS
		case 'a':
		case 'A':
			/* RLIMIT_AS - max address space (KB) */
			retval |= setrlimit_value (RLIMIT_AS, pp, 1024);
			break;
#endif
#ifdef RLIMIT_CPU
		case 't':
		case 'T':
			/* RLIMIT_CPU - max CPU time (MIN) */
			retval |= setrlimit_value (RLIMIT_CPU, pp, 60);
			break;
#endif
#ifdef RLIMIT_DATA
		case 'd':
		case 'D':
			/* RLIMIT_DATA - max data size (KB) */
			retval |= setrlimit_value (RLIMIT_DATA, pp, 1024);
			break;
#endif
#ifdef RLIMIT_FSIZE
		case 'f':
		case 'F':
			/* RLIMIT_FSIZE - Maximum filesize (KB) */
			retval |= setrlimit_value (RLIMIT_FSIZE, pp, 1024);
			break;
#endif
#ifdef RLIMIT_NPROC
		case 'u':
		case 'U':
			/* RLIMIT_NPROC - max number of processes */
			retval |= setrlimit_value (RLIMIT_NPROC, pp, 1);
			break;
#endif
#ifdef RLIMIT_CORE
		case 'c':
		case 'C':
			/* RLIMIT_CORE - max core file size (KB) */
			retval |= setrlimit_value (RLIMIT_CORE, pp, 1024);
			break;
#endif
#ifdef RLIMIT_MEMLOCK
		case 'm':
		case 'M':
			/* RLIMIT_MEMLOCK - max locked-in-memory address space (KB) */
			retval |= setrlimit_value (RLIMIT_MEMLOCK, pp, 1024);
			break;
#endif
#ifdef RLIMIT_NOFILE
		case 'n':
		case 'N':
			/* RLIMIT_NOFILE - max number of open files */
			retval |= setrlimit_value (RLIMIT_NOFILE, pp, 1);
			break;
#endif
#ifdef RLIMIT_RSS
		case 'r':
		case 'R':
			/* RLIMIT_RSS - max resident set size (KB) */
			retval |= setrlimit_value (RLIMIT_RSS, pp, 1024);
			break;
#endif
#ifdef RLIMIT_STACK
		case 's':
		case 'S':
			/* RLIMIT_STACK - max stack size (KB) */
			retval |= setrlimit_value (RLIMIT_STACK, pp, 1024);
			break;
#endif
		case 'k':
		case 'K':
			retval |= set_umask (pp);
			break;
		case 'l':
		case 'L':
			/* LIMIT the number of concurent logins */
			retval |= check_logins (name, pp);
			break;
		case 'p':
		case 'P':
			retval |= set_prio (pp);
			break;
		}
	return retval;
}

static int setup_user_limits (const char *uname)
{
	/* TODO: allow and use @group syntax --cristiang */
	FILE *fil;
	char buf[1024];
	char name[1024];
	char limits[1024];
	char deflimits[1024];
	char tempbuf[1024];

	/* init things */
	memzero (buf, sizeof (buf));
	memzero (name, sizeof (name));
	memzero (limits, sizeof (limits));
	memzero (deflimits, sizeof (deflimits));
	memzero (tempbuf, sizeof (tempbuf));

	/* start the checks */
	fil = fopen (LIMITS_FILE, "r");
	if (fil == NULL) {
		return 0;
	}
	/* The limits file have the following format:
	 * - '#' (comment) chars only as first chars on a line;
	 * - username must start on first column
	 * A better (smarter) checking should be done --cristiang */
	while (fgets (buf, 1024, fil) != NULL) {
		if (buf[0] == '#' || buf[0] == '\n')
			continue;
		memzero (tempbuf, sizeof (tempbuf));
		/* a valid line should have a username, then spaces,
		 * then limits
		 * we allow the format:
		 * username    L2  D2048  R4096
		 * where spaces={' ',\t}. Also, we reject invalid limits.
		 * Imposing a limit should be done with care, so a wrong
		 * entry means no care anyway :-). A '-' as a limits
		 * strings means no limits --cristiang */
		if (sscanf (buf, "%s%[ACDFMNRSTULPacdfmnrstulp0-9 \t-]",
			    name, tempbuf) == 2) {
			if (strcmp (name, uname) == 0) {
				strcpy (limits, tempbuf);
				break;
			} else if (strcmp (name, "*") == 0) {
				strcpy (deflimits, tempbuf);
			}
		}
	}
	fclose (fil);
	if (limits[0] == '\0') {
		/* no user specific limits */
		if (deflimits[0] == '\0')	/* no default limits */
			return 0;
		strcpy (limits, deflimits);	/* use the default limits */
	}
	return do_user_limits (limits, uname);
}
#endif				/* LIMITS */


static void setup_usergroups (const struct passwd *info)
{
	const struct group *grp;
	mode_t oldmask;

/*
 *	if not root, and UID == GID, and username is the same as primary
 *	group name, set umask group bits to be the same as owner bits
 *	(examples: 022 -> 002, 077 -> 007).
 */
	if (info->pw_uid != 0 && info->pw_uid == info->pw_gid) {
		grp = getgrgid (info->pw_gid);
		if (grp && (strcmp (info->pw_name, grp->gr_name) == 0)) {
			oldmask = umask (0777);
			umask ((oldmask & ~070) | ((oldmask >> 3) & 070));
		}
	}
}

/*
 *	set the process nice, ulimit, and umask from the password file entry
 */

void setup_limits (const struct passwd *info)
{
	char *cp;
	int i;
	long l;

	if (getdef_bool ("USERGROUPS_ENAB"))
		setup_usergroups (info);

	/*
	 * See if the GECOS field contains values for NICE, UMASK or ULIMIT.
	 * If this feature is enabled in /etc/login.defs, we make those
	 * values the defaults for this login session.
	 */

	if (getdef_bool ("QUOTAS_ENAB")) {
#ifdef LIMITS
		if (info->pw_uid != 0)
			if (setup_user_limits (info->pw_name) &
			    LOGIN_ERROR_LOGIN) {
				fprintf (stderr, _("Too many logins.\n"));
				sleep (2);
				exit (1);
			}
#endif
		for (cp = info->pw_gecos; cp != NULL; cp = strchr (cp, ',')) {
			if (*cp == ',')
				cp++;

			if (strncmp (cp, "pri=", 4) == 0) {
				i = atoi (cp + 4);
				if (i >= -20 && i <= 20)
					(void) nice (i);

				continue;
			}
			if (strncmp (cp, "ulimit=", 7) == 0) {
				l = strtol (cp + 7, (char **) 0, 10);
				set_filesize_limit (l);
				continue;
			}
			if (strncmp (cp, "umask=", 6) == 0) {
				i = strtol (cp + 6, (char **) 0, 8) & 0777;
				(void) umask (i);

				continue;
			}
		}
	}
}

#endif				/* !USE_PAM */
